File: src/repositories/interfaces/auth.repository.ts
import { kUserNotFound } from "../../common/failures";
import { Failure } from "../../common/types";
import { IAuthEntity, IRefreshSessionEntity } from "../../entities";

export interface IAuthRepository {
  getCognitoId(accessToken: string): Promise<string | Failure>;

  refreshSession(
    refreshToken: string,
    email: string
  ): Promise<IRefreshSessionEntity | Failure>;

  deleteUser(cognitoId: string): Promise<void | Failure>;

  registerUser(email: string, password: string): Promise<string | Failure>;

  signOut(accessToken: string): Promise<void | Failure>;

  forgotPassword(email: string): Promise<void | Failure>;

  forgotPasswordSubmit(
    email: string,
    code: string,
    newPassword: string
  ): Promise<void | Failure>;

  confirmRegister(email: string, code: string): Promise<void | Failure>;

  login(email: string, password: string): Promise<IAuthEntity | Failure>;

  resendConfirmationCode(email: string): Promise<void | Failure>;
}

File: lib/src/repositories/interfaces/auth.repository.dart
import 'package:dartz/dartz.dart';

import '../../entities/entities.dart';

abstract class IAuthRepository {
  Future<Either<AuthTokensEntity, Failure>> login(
    String email,
    String password, {
    Duration? timeout,
  });

  Future<Failure?> register(
    String email,
    String password,
    String gymId,
    String firstName,
    String lastName,
  );

  Future<Either<SessionRefreshTokensEntity, Failure>> refreshSession(
    String email,
    String refreshToken, {
    Duration? timeout,
  });

  Future<bool> validAccessToken(String accessToken);

  Future<Failure?> confirmRegister(String email, String code);

  Future<Failure?> resendConfirmation(String email);

  Future<Failure?> forgotPassword(String email);

  Future<Failure?> resetPassword(
    String email,
    String code,
    String newPassword,
  );
}
